### PR TITLE
Iterative fitting with outlier removal

### PR DESCRIPTION
**Pull request by mirca**
Hi all,

this PR addresses #4698.

The basic idea is (as stated by @pllim):
1. Create a model.
2. Attach a fitter to the model.
3. Fit with some data.
4. Reject some outliers with sigma clipping (or any other outlier removal procedure).
5. Redo fitting with clipped data.
6. Repeat 4-5 for a number of times (niter) until satisfied.

So, I would like to know what you guys think...

Can I go ahead and write some tests for the rest of the fitters and consider 2D case as well?

See the example below as a use case:
```
import numpy as np
from astropy.stats import sigma_clip
from astropy.modeling import models, fitting
import scipy.stats as stats
from matplotlib import pyplot as plt

np.random.seed(0)
x = np.linspace(-5., 5., 200)
y = 3 * np.exp(-0.5 * (x - 1.3)**2 / 0.8**2)
c = stats.bernoulli.rvs(0.25, size=x.shape)
y += np.random.normal(0., 0.2, x.shape) + c*np.random.normal(3.0, 5.0, x.shape)

g_init = models.Gaussian1D(amplitude=1., mean=0, stddev=1.)
fit = fitting.FittingWithOutlierRemoval(fitting.LevMarLSQFitter(),
                                        sigma_clip, niter=3, sigma=3.0)
filtered_data, fitted_model = fit(g_init, x, y)

plt.figure()
plt.plot(x, y, 'gx', label="data")
plt.plot(x[~filtered_data.mask], filtered_data.data[~filtered_data.mask],
         'b+', label="filtered data")
plt.plot(x, fitted_model(x), '--', label="fitted model")
plt.legend(loc=2)
plt.show()
```
figure_1

Thanks!

### CODE REVIEW

**Comment by pllim** [comment]
@hcferguson , do you have test cases you wish to use here? c/c @larrybradley , @nden , @astrofrog

**Comment by crawfordsm** [comment]
One additional feature that would be nice here would be to add in a tolerance parameter so that it converges once the residual is below some value.

**Comment by crawfordsm** [comment]
Also adding in the example to the documentation would be helpful or as a separate page in the astropy documents

**Comment by hcferguson** [comment]
@plim Nothing handy where I have results already at hand from doing this in IRAF or my own python scripts. 

A common 2-D use case would be trying to fit sky background in an image that still has residuals from poorly masked sources. To apply the figure-of-merit that we developed in the background-fitting sprint, I would suggest creating an artificial image with a gradient, adding point sources and noise, filtering out the point sources using photutils, surface-fitting the gradient, and then comparing the rms as a function of scale to the pure-noise case. Fiddle with the point-sources and masking so that it doesn't do very well on the first pass and see if iterative rejection helps.

**Comment by pllim** [comment]
This looks promising. But I am not sure how well this goes with the modeling/fitting discussions at the recent Python in Astronomy 2016. There were talks about using external fitters from Sherpa instead?

**Comment by mirca** [comment]
@pllim I was not at PyAstro'16, but I'm curious about what thoughts were discussed regarding ``astropy.modeling`` and ``astropy.stats`` primarily :)

**Comment by pllim** [comment]
@mirca , I didn't participate in those discussions myself, so I was hoping @nden, @astrofrog, etc. who did could elaborate.

**Comment by hamogu** [comment]
I was not at the discussion either, but given that I brought up astropy<->sherpa as a GSoC project, I;ve thought about this quite a bit:
I don't think that Sherpa fitters should *replace* astropy fitters, they should be offered *in addition* since sherpa offers some methods that astropy does not. The fitter in this PR is currently not available in either package, so I'm definitely looking forward to have that go into astropy.

**Comment by mirca** [comment]
An example for fitting a two-dimensional Gaussian case is:

```
import numpy as np
from astropy.stats import sigma_clip
from astropy.modeling import models, fitting
import scipy.stats as stats
import matplotlib
from matplotlib import pyplot as plt

def initial_guess_Gaussian2D(data, grid):
    y = grid[0]
    x = grid[1]
    
    wx = x*data
    wy = y*data
    total_intensity = np.sum(data)

    x_mean = np.sum(wx)/total_intensity
    y_mean = np.sum(wy)/total_intensity

    x_to_pixel = x[0].size/(x[x[0].size-1][x[0].size-1] - x[0][0])
    y_to_pixel = y[0].size/(y[y[0].size-1][y[0].size-1] - y[0][0])
    x_pos = np.around(x_mean*x_to_pixel + x[0].size/2.).astype(int)
    y_pos = np.around(y_mean*y_to_pixel + y[0].size/2.).astype(int)
    amplitude = data[y_pos][x_pos]

    return amplitude, x_mean, y_mean

np.random.seed(0)
y, x = np.mgrid[-3:3:128j, -3:3:128j]
z = 3 * np.exp(-0.5*(x-1.0)**2/0.8**2 - 0.5*y**2/0.8**2)
c = stats.bernoulli.rvs(0.25, size=z.shape)
data = z + np.random.normal(0., 0.2, z.shape) + c*np.random.normal(z, 2.0, z.shape)
grid = np.array([y,x])
guess = initial_guess_Gaussian2D(data, grid)
g2_init = models.Gaussian2D(amplitude=guess[0], x_mean=guess[1],
                            y_mean=guess[2], x_stddev=.75, y_stddev=1.25)
ORfit = fitting.FittingWithOutlierRemoval(fitting.LevMarLSQFitter(),
                                          sigma_clip, niter=3, sigma=4.0)
filtered_data, fitted_model = ORfit(g2_init, x, y, data)

plt.figure(figsize=(16, 5))
plt.subplot(1, 6, 1)
plt.imshow(z, cmap='magma')
plt.title("Original model")
plt.subplot(1, 6, 2)
plt.imshow(data, cmap='magma')
plt.title("Synthetic data")
plt.subplot(1, 6, 3)
plt.imshow(g2_init(x,y), cmap='magma')
plt.title("Initial guess model")
plt.subplot(1, 6, 4)
plt.imshow(filtered_data, cmap='magma')
plt.title("Sigma clipped data")
plt.subplot(1, 6, 5)
plt.imshow(fitted_model(x, y), cmap='magma')
plt.title("Fitted model")
plt.subplot(1, 6, 6)
plt.imshow(data - fitted_model(x, y), cmap='magma')
plt.title("Residual")
plt.show()
```
figure_1

Also, for the sake of completeness, the estimates for each step are:
```
                  amplitude     x_mean      y_mean     x_stddev   y_stddev   theta
                ------------- ----------- ----------  ----------  ---------  -----
(true params)   [ 3.0         1.0         0.0         0.8         0.8         0.0]
(initial guess) [ 3.18157654  1.0622778   0.06024728  0.75        1.25        0.0]
(no sigma clip) [ 3.87895303  1.01590443  0.02267679  0.76910491  0.79378423  0.59149587]
(1st iteration) [ 3.0731814   1.01265775  0.0061559   0.81723209  0.83768409  0.30759714]
(2nd iteration) [ 3.06989421  1.01206323  0.00622122  0.81737112  0.8382779   0.29975979]
(3rd iteration) [ 3.06988456  1.01206267  0.00622002  0.81737271  0.8382812   0.29971142]
```

@hcferguson Thanks for the suggestion! I'll work on it soon.
@hamogu Nice to hear this. :)

**Comment by eteq** [comment]
Just to give a bit more background on the discussion above (I *was* at pyastro16): there was some consideration that we might want to deprecate `astropy.fitting` or move it to a separate affiliated package based on the idea that it is *one* way to do fitting, and there are many others that are arguably more rigorous.  So `astropy.fitting` perhaps should not get special status as sort of a "reference implementation".  Instead it should be considered "one of many" and we should put more effort into getting more (external?) fitters being compatible with `astropy.models`.

No firm decision was reached, though - some of the meeting attendees are considering drawing up an APE about it.

**Comment by mirca** [comment]
Any thoughts here? :)

**Comment by hamogu** [comment]
@eteq the fact that somebody might draw up an APE in the future should not stop is to work on this PR now. If fitting is moved to ours own package, out can be moved including this merged PR.
So, are there any concerns with the content of this PR as opposed to concerns with the future of the package out resides in?

**Comment by hamogu** [comment]
@mirca can you add a change log entry?

**Comment by nden** [comment]
I agree that a potential move of `astropy.fitting` is not a reason to stop improving it. I'll try to review this soon.

File: astropy/modeling/fitting.py [hunk 1]
```diff
@@ -341,6 +341,118 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+        ----------
+        fitter : An Astropy fitter
+            An instance of any Astropy fitter, i.e., LinearLSQFitter,
+            LevMarLSQFitter, SLSQPLSQFitter, SimplexLSQFitter, JointFitter.
+        outlier_func : function
+            A function for outlier removal.
+        niter : int (optional)
+            Number of iterations.
+        outlier_kwargs : dict (optional)
+            Keyword arguments for outlier_func.
```
Comment (nden) [inline astropy/modeling/fitting.py]: Move the `Parameters` section to the class docstrings, just like the rest of the classes.

File: astropy/modeling/fitting.py [hunk 2]
```diff
@@ -341,6 +341,118 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+        """
+
+        self.fitter = fitter
+        self.outlier_func = outlier_func
+        self.niter = niter
```
Comment (nden) [inline astropy/modeling/fitting.py]: Make the three above private, e.g. `self._niter = niter`.

File: astropy/modeling/fitting.py [hunk 3]
```diff
@@ -341,6 +341,118 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+    @niter.setter
+    def niter(self, niter):
+        if isinstance(niter, type(1)):
+            self._niter = niter
+        else:
+            raise ValueError('niter is expected to be an integer.')
+
+    def __str__(self):
+        return ("Fitter: {0}\nOutlier function: {1}\nNum. of iterations: {2}" +
+                ("\nOutlier func. args.: {3}")).format(self.fitter,
```
Comment (nden) [inline astropy/modeling/fitting.py]: Would it be better to print the name, `self.fitter.__class__.__name__`.

File: astropy/modeling/fitting.py [hunk 4]
```diff
@@ -341,6 +341,118 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+                                                       self.outlier_func,
```
Comment (nden) [inline astropy/modeling/fitting.py]: Similarly `self.outlier_func.__name__`

File: astropy/modeling/fitting.py [hunk 5]
```diff
@@ -341,6 +341,118 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+                                                       self.niter,
+                                                       self.outlier_kwargs)
+
+    def __call__(self, model, x, y, z=None, weights=None):
```
Comment (nden) [inline astropy/modeling/fitting.py]: Does it make sense to add here `**kwargs` to pass optional arguments to the fitter?

File: astropy/modeling/fitting.py [hunk 6]
```diff
@@ -341,6 +341,118 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+
+        fitted_model = self.fitter(model, x, y, z, weights)
+        if z is None:
+            filtered_data = y
+            for n in range(self.niter):
+                filtered_data = self.outlier_func(filtered_data,
+                                                  **self.outlier_kwargs)
+                fitted_model = self.fitter(fitted_model,
+                                           x[~filtered_data.mask],
+                                           filtered_data.data[~filtered_data.mask])
```
Comment (nden) [inline astropy/modeling/fitting.py]: Add `**kwargs` to the fitter call here.

File: astropy/modeling/tests/test_fitters.py [hunk 7]
```diff
@@ -403,3 +404,110 @@ def test_param_cov(self):
...
+        self.x = np.linspace(-5., 5., 200)
+        self.model_params = (3.0, 1.3, 0.8)
+
+        def func(p, x):
+            return p[0]*np.exp(-0.5*(x - p[1])**2/p[2]**2)
+
+        self.y = func(self.model_params, self.x)
+
+    def test_with_fitters_and_sigma_clip(self):
+        from astropy.stats import sigma_clip
```
Comment (nden) [inline astropy/modeling/tests/test_fitters.py]: Move this import to the top of the file and write it as `from ...stats import sigma_clip`.

File: astropy/modeling/tests/test_fitters.py [hunk 8]
```diff
@@ -403,3 +404,110 @@ def test_param_cov(self):
...
+        import scipy.stats as stats
+        from astropy.modeling import fitting
```
Comment (nden) [inline astropy/modeling/tests/test_fitters.py]: This line is not needed. There's `from ..fitting import *` so the class name should be already available.

File: astropy/modeling/tests/test_fitters.py [hunk 9]
```diff
@@ -403,3 +404,110 @@ def test_param_cov(self):
...
+        y_to_pixel = y[0].size/(y[y[0].size-1][y[0].size-1] - y[0][0])
+        x_pos = np.around(x_mean*x_to_pixel + x[0].size/2.).astype(int)
+        y_pos = np.around(y_mean*y_to_pixel + y[0].size/2.).astype(int)
+
+        amplitude = data[y_pos][x_pos]
+
+        return amplitude, x_mean, y_mean
+
+    def test_with_fitters_and_sigma_clip(self):
+        from astropy.stats import sigma_clip
```
Comment (nden) [inline astropy/modeling/tests/test_fitters.py]: Remove this line.

**Comment by nden** [comment]
@mirca Overall I like this. I left some comments inline. In addition, please add a `__repr__` and run this through a PEP8 checking program.

File: docs/modeling/fitting.rst [hunk 10]
```diff
@@ -60,6 +60,46 @@ Fitting examples
...
+     # Generate fake data with outliers
+     np.random.seed(0)
+     x = np.linspace(-5., 5., 200)
+     y = 3 * np.exp(-0.5 * (x - 1.3)**2 / 0.8**2)
+     c = stats.bernoulli.rvs(0.35, size=x.shape)
+     y += (np.random.normal(0., 0.2, x.shape) +
+           c*np.random.normal(3.0, 5.0, x.shape))
+     g_init = models.Gaussian1D(amplitude=1., mean=0, stddev=1.)
+
+     # initiliaze fitters
```
Comment (MSeifert04) [inline docs/modeling/fitting.rst]: "initialize"

Comment (mirca) [inline docs/modeling/fitting.rst]: Thanks!

File: astropy/modeling/fitting.py (hunk 1 above)
Comment (mirca) [inline astropy/modeling/fitting.py]: Done.

File: astropy/modeling/fitting.py (hunk 2 above)
Comment (mirca) [inline astropy/modeling/fitting.py]: Can I leave them as is? I'm using properties here, so I think this is safe :)

File: astropy/modeling/fitting.py (hunk 5 above)
Comment (mirca) [inline astropy/modeling/fitting.py]: Totally! Thanks

**Comment by mirca** [comment]
@nden Thanks a bunch! Does the changes look good? How do you like the `__repr__()` ?

**Comment by mirca** [comment]
@nden any thoughts?

File: astropy/modeling/fitting.py [hunk 11]
```diff
@@ -341,6 +341,129 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
         return model_copy
 
 
+class FittingWithOutlierRemoval(object):
+    """
+    This class combines an outlier removal technique with a fitting procedure.
+    Basically, given a number of iterations `niter`, outliers are removed in
+    and fitting is performed for each iteration.
+
+    Attributes
```
Comment (nden) [inline astropy/modeling/fitting.py]: Please `Parameters`, not `Attributes` as in the rest of the astropy docstrings.

File: astropy/modeling/fitting.py [hunk 12]
```diff
@@ -341,6 +341,129 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+
+    def __init__(self, fitter, outlier_func, niter=3, **outlier_kwargs):
+        self.fitter = fitter
+        self.outlier_func = outlier_func
+        self.niter = niter
+        self.outlier_kwargs = outlier_kwargs
+
+    @property
+    def fitter(self):
+        return self._fitter
```
Comment (nden) [inline astropy/modeling/fitting.py]: Here you return `self._fitter` but the attribute above is `self.fitter`. I think the property is ignored if this works. If you want to use `self.fitter` simply remove the `fitter` property. Similarly for the other attributes.

File: astropy/modeling/fitting.py [hunk 13]
```diff
@@ -341,6 +341,129 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
...
+                               x[~filtered_data.mask],\
+                               filtered_data.data[~filtered_data.mask],\
+                               **kwargs)
+        else:
+            filtered_data = z
+            for n in range(self.niter):
+                filtered_data = self.outlier_func(filtered_data,
+                                                  **self.outlier_kwargs)
+                fitted_model = self.fitter(fitted_model,\
+                               x[~filtered_data.mask],\
```
Comment (nden) [inline astropy/modeling/fitting.py]: back slash is not necessary in this case.

File: astropy/modeling/tests/test_fitters.py [hunk 14]
```diff
@@ -403,3 +405,107 @@ def test_param_cov(self):
...
+        center position"""
+
+        wx = x*data
+        wy = y*data
+        total_intensity = np.sum(data)
+        x_mean = np.sum(wx)/total_intensity
+        y_mean = np.sum(wy)/total_intensity
+
+        x_to_pixel = x[0].size/(x[x[0].size-1][x[0].size-1] - x[0][0])
+        y_to_pixel = y[0].size/(y[y[0].size-1][y[0].size-1] - y[0][0])
```
Comment (nden) [inline astropy/modeling/tests/test_fitters.py]: add spaces around the minus sign in the above two lines.

**Comment by nden** [comment]
@mirca Sorry for the long silence. I added a few minor comments. 
When you are done could you edit the top description to remove  "work in progress" .
Thanks.

**Comment by mirca** [comment]
@nden done!

**Comment by nden** [comment]
Thanks @mirca!

**Pull request merged**